# Schritt 1 
1.1: Ermitteln des umzusystematisierenden Bestands mit SQL-Befehl, Ausgabe: csv-Datei  
1.2: sortiert nach Spalten (Signaturen, Mediennummern, Kurztitel, Autor, Reihentitel, ISBN, neue Notation, neue Schlagwörter, neue Gesamtsignatur)  
1.3: Signaturen numerisch aufsteigend sortieren mit zuvor geschriebenem Makro (statt P1, P10, P100 > P1, P2, P3 ...)  

In [1]:
import urllib.request
import pandas as pd

In [4]:
# Set the source url
url = "https://raw.githubusercontent.com/adalka/Malis18_T9.2_Automatisierte_Schlagwort_RVK-Notationsvergabe/master/2019-05-27_P1000_sortiert_mit_ISBN_Exnr_als_Text.csv?token=AL2CQ2TCLBZN2N65SWQDKM25CCRJQ"
apc_file = "2019-05-27_P1000_sortiert_mit_ISBN_Exnr_als_Text.csv"

In [5]:
urllib.request.urlretrieve(url, apc_file)

('2019-05-27_P1000_sortiert_mit_ISBN_Exnr_als_Text.csv',
 <http.client.HTTPMessage at 0x1d58caec8d0>)

In [55]:
# Datei war zu groß "We can't make this file beautiful and searchable because it's too large."
# Set the new source url
# Datei enthält Fehler "We can make this file beautiful and searchable if this error is corrected: Illegal quoting in line 4"
# We can make this file beautiful and searchable if this error is corrected: It looks like row 2 should actually have 1 column, instead of 2. in line 1."
# hilfreich: https://gist.github.com/jexp/2faa74b983907b43586d
# hilfreich: https://help.github.com/en/articles/rendering-csv-and-tsv-data

# Schritt 2
2.1: csv-Datei aufs Wesentliche reduziert: Titelnummern und ISBN

In [57]:
# csv-Datei war nicht durch Komma, sondern Semikolon getrennt, öffnen im Notepad > Semikolons ersetzt durch Kommas

In [18]:
apcs = pd.read_csv(apc_file)

In [19]:
apcs

EXEMPLARNR           ISBN
0    1.586749e+09          FALSE
1    1.586750e+09          FALSE
2    1.586751e+09          FALSE
3    1.586752e+09          FALSE
4    1.586754e+09  3-85460-115-8
5    1.586755e+09  3-85460-111-5
6    2.136820e+08          FALSE
7    1.586757e+09  91-7192-927-4
8    1.586758e+09  3-88609-200-3
9    1.586759e+09  0-521-23742-4
10   1.586765e+09  3-496-02548-4
11   1.586768e+09  3-496-02550-6
12   1.586932e+09          FALSE
13   5.032740e+08  3-87065-560-7
14   1.586936e+09  3-8053-1537-6
15   1.586942e+09  91-506-1093-7
16   1.586943e+09  91-506-1084-8
17   1.586944e+09          FALSE
18   1.586946e+09          FALSE
19   1.586949e+09          FALSE
20   1.586952e+09          FALSE
21   1.586954e+09          FALSE
22   1.586958e+09          FALSE
23   2.210330e+08  3-923166-43-5
24   4.078390e+08          FALSE
25   1.586962e+09          FALSE
26   1.586966e+09          FALSE
27   1.586968e+09          FALSE
28   1.586974e+09          FALSE
29   1.586978e+09  3-8067-0200-4
..            ...            ...
970  2.157710e+08          FALSE
971  2.157700e+08          FALSE
972  1.590649e+09          FALSE
973  1.590652e+09          FALSE
974  1.590654e+09          FALSE
975  1.590660e+09  3-406-06775-1
976  1.590665e+09          FALSE
977  1.590667e+09          FALSE
978  1.590669e+09          FALSE
979  1.590670e+09          FALSE
980  1.590671e+09          FALSE
981  1.590674e+09          FALSE
982  1.590676e+09          FALSE
983  1.590677e+09          FALSE
984  1.590686e+09          FALSE
985  1.590687e+09          FALSE
986  1.590688e+09          FALSE
987  1.590689e+09  3-929055-36-8
988  1.590690e+09          FALSE
989  1.590691e+09          FALSE
990  1.590693e+09          FALSE
991  1.590694e+09          FALSE
992  1.590695e+09          FALSE
993  1.590696e+09          FALSE
994  1.590697e+09          FALSE
995  1.590698e+09          FALSE
996  1.590699e+09          FALSE
997  1.590700e+09          FALSE
998  1.590701e+09          FALSE
999  1.590702e+09          FALSE

[1000 rows x 2 columns]

In [15]:
print(apcs)

       EXEMPLARNR           ISBN
0    1.586749e+09          FALSE
1    1.586750e+09          FALSE
2    1.586751e+09          FALSE
3    1.586752e+09          FALSE
4    1.586754e+09  3-85460-115-8
5    1.586755e+09  3-85460-111-5
6    2.136820e+08          FALSE
7    1.586757e+09  91-7192-927-4
8    1.586758e+09  3-88609-200-3
9    1.586759e+09  0-521-23742-4
10   1.586765e+09  3-496-02548-4
11   1.586768e+09  3-496-02550-6
12   1.586932e+09          FALSE
13   5.032740e+08  3-87065-560-7
14   1.586936e+09  3-8053-1537-6
15   1.586942e+09  91-506-1093-7
16   1.586943e+09  91-506-1084-8
17   1.586944e+09          FALSE
18   1.586946e+09          FALSE
19   1.586949e+09          FALSE
20   1.586952e+09          FALSE
21   1.586954e+09          FALSE
22   1.586958e+09          FALSE
23   2.210330e+08  3-923166-43-5
24   4.078390e+08          FALSE
25   1.586962e+09          FALSE
26   1.586966e+09          FALSE
27   1.586968e+09          FALSE
28   1.586974e+09          FALSE
29   1.586

In [18]:
# Spalten anzeigen
apcs.columns

Index(['EXEMPLARNR', 'ISBN'], dtype='object')

In [19]:
# nur ISBN anzeigen
apcs["ISBN"]

0              FALSE
1              FALSE
2              FALSE
3              FALSE
4      3-85460-115-8
5      3-85460-111-5
6              FALSE
7      91-7192-927-4
8      3-88609-200-3
9      0-521-23742-4
10     3-496-02548-4
11     3-496-02550-6
12             FALSE
13     3-87065-560-7
14     3-8053-1537-6
15     91-506-1093-7
16     91-506-1084-8
17             FALSE
18             FALSE
19             FALSE
20             FALSE
21             FALSE
22             FALSE
23     3-923166-43-5
24             FALSE
25             FALSE
26             FALSE
27             FALSE
28             FALSE
29     3-8067-0200-4
           ...      
970            FALSE
971            FALSE
972            FALSE
973            FALSE
974            FALSE
975    3-406-06775-1
976            FALSE
977            FALSE
978            FALSE
979            FALSE
980            FALSE
981            FALSE
982            FALSE
983            FALSE
984            FALSE
985            FALSE
986          

# Schritt 3
3.1: csv-Datei korrigieren: Exemplarnr. und ISBN als Text  
3.2: in csv-Datei alle leeren Zeilen in Spalte "ISBN" in "FALSE" benannt  
3.3: alle Semikolons ersetzt durch Kommas  
3.4: csv-Datei war nicht durch Komma, sondern Semikolon getrennt, öffnen im Notepad > Semikolons ersetzt durch Kommas  
3.5: leere Zeilen mit FALSE ausgefüllt: Markieren des Zellbereichs, dessen Leerzellen ausgefüllt werden sollen; Befehl GEHE ZU (Tastenkombination STRG G). Schaltfläche INHALTE, Aktivieren von LEERZELLEN 
3.6: Gleichheitszeichen eingeben, Taste PFEIL OBEN, STRG EINGABE  
3.7: alle leeren Zeilen werden mit "FALSE" ausgefüllt, sodass man weiß, dass diese keine ISBN enthalten  

# Schritt 4
4.1: alle obigen url.lib request wiederholt, durchlaufen lassen  
4.2: apcs werden leider nicht als "Text", sondern als "Zahl" dargestellt (EXEMPLARNR muss aber lesbar bleiben, da sie dem genauen Titel zugewiesen wird)    
4.3: durch folgenden import behoben:  

In [ ]:
import csv
import csv
with open('2019-05-27_P1000_sortiert_mit_ISBN_Exnr_als_Text.csv', newline='') as f:
    reader = csv.reader(f)
    for row in reader:
        print(row)
# nicht ausgeführt, da Datei zu groß (1000 Zeilen)
# Ausführen ergibt aber Zahlen statt Text, d. h. EXEMPLARNR werden richtig dargestellt!

# Schritt 5 (Exkurs): Eingabe von ISBN in google spreadsheet  
5.1: Es gibt ein Google Spreadsheet, in dem man ISBN testweise eingeben kann: https://docs.google.com/spreadsheets/d/1hZ92Pxgw2hbiTVmTG0XAGNMPYzN6wmvm-mrLPvZ-eoM/edit#gid=0  
5.2: dies ermöglicht die Ausgabe von RVK-Notationen für eingegebene ISBN  
5.3: RVK-Notationen können aber auch importiert werden, z.B. =ImportJSON("http://data.bib.uni-mannheim.de/malibu/isbn/swb.php?isbn="ISBN"&format=json"; "/rvk"; "noHeaders")  
5.4: Befehl ausführen: JSON-basierte Antwortdaten (s. https://2.python-requests.org//de/latest/user/quickstart.html)  

In [20]:
import requests
r = requests.get('http://data.bib.uni-mannheim.de/malibu/isbn/gbv.php?isbn=3-85460-115-8&format=json')
r.json()

{'id': ['316021121'],
 'isbn': ['3854601158'],
 'dnbNr': '',
 'oclcNr': ['165236530'],
 'titel': ['Sie sahen die Mauern Trojas'],
 'untertitel': ['Kunstgewerbe der Bronzezeit in Südostungarn : Sonderausstellung im Museum für Urgeschichte, Aspern/Zaya, 1. April - 15. Nov. 1994'],
 'autor': ['Otto Trogmayer, Gabriella Vörös'],
 'gesamttitel': ['Katalog des Niederösterreichischen Landesmuseums N.F. 348'],
 'hochschulvermerk': '',
 'auflage': '',
 'erscheinungsinfo': ['Aspern/Zaya', 'Museum für Urgeschichte', '1994'],
 'umfang': ['34 Seiten', 'Illustrationen'],
 'links': '',
 'bestand': ['ZK Sachsen Anhalt',
  'SUB+Uni Hamburg <18>',
  'ULB Halle <3>',
  'LDA Halle <Ha 91>',
  'PKB (Museen) Berlin',
  'Museen+Denkmalschutzamt HH'],
 'rvk': ['NF 1665'],
 'ddc': '',
 'sw': [],
 'produktSigel': '',
 'einzelaufnahmen': [{'id': ['316021121'],
   'isbn': ['3854601158'],
   'dnbNr': '',
   'oclcNr': ['165236530'],
   'titel': ['Sie sahen die Mauern Trojas'],
   'untertitel': ['Kunstgewerbe der Br

# Schritt 6: RVK-Notationen auslesen  
6.1: Die RVK-Notationen (= Werte) befinden sich unter dem Key 'rvk' ('rvk': ['NF 1665'])  
6.2: nächster Schritt: Notationen auslesen lassen für ISBN  

In [40]:
import requests

In [41]:
r = requests.get('http://data.bib.uni-mannheim.de/malibu/isbn/gbv.php?isbn=3-85460-115-8&format=json')

In [42]:
r.text

'{\n    "id": [\n        "316021121"\n    ],\n    "isbn": [\n        "3854601158"\n    ],\n    "dnbNr": "",\n    "oclcNr": [\n        "165236530"\n    ],\n    "titel": [\n        "Sie sahen die Mauern Trojas"\n    ],\n    "untertitel": [\n        "Kunstgewerbe der Bronzezeit in S\\u00fcdostungarn : Sonderausstellung im Museum f\\u00fcr Urgeschichte, Aspern\\/Zaya, 1. April - 15. Nov. 1994"\n    ],\n    "autor": [\n        "Otto Trogmayer, Gabriella V\\u00f6r\\u00f6s"\n    ],\n    "gesamttitel": [\n        "Katalog des Nieder\\u00f6sterreichischen Landesmuseums N.F. 348"\n    ],\n    "hochschulvermerk": "",\n    "auflage": "",\n    "erscheinungsinfo": [\n        "Aspern\\/Zaya",\n        "Museum f\\u00fcr Urgeschichte",\n        "1994"\n    ],\n    "umfang": [\n        "34 Seiten",\n        "Illustrationen"\n    ],\n    "links": "",\n    "bestand": [\n        "ZK Sachsen Anhalt",\n        "SUB+Uni Hamburg <18>",\n        "ULB Halle <3>",\n        "LDA Halle <Ha 91>",\n        "PKB (Muse

Key und Value lauten "rvk": [\n        "NF 1665"\n    ]  
Wie erhalte ich eine automatische Eingabe der Spalte ISBN in den Query String? (s. http://www.compciv.org/guides/python/how-tos/creating-proper-url-query-strings/)  

# Schritt 7: ISBN in Query String setzen    
7.1: Wenn die ISBN aus der Spalte "ISBN" der csv in den Query String gesetzt wird, können Abfragen für alle in der Liste vorhandenen ISBN druchgeführt werden  
7.2: Wie können Werte der Spalte "ISBN" in den String gebracht werden: 'http://data.bib.uni-mannheim.de/malibu/isbn/gbv.php?isbn=3-85460-115-8&format=json'   

In [43]:
from urllib.request import urlretrieve

In [16]:
thing = urlretrieve("http://data.bib.uni-mannheim.de/malibu/isbn/gbv.php?isbn=978-3-8006-4582-4&format=json")

In [17]:
from urllib.parse import quote

In [19]:
qstr = {3-85460-115-8}
# hier müsste die Spalte "ISBN" aus der csv definiert werden
thing = urlretrieve("http://data.bib.uni-mannheim.de/malibu/isbn/gbv.php?isbn=" + qstr + "&format=json")
# so müsste der String danach aufgebaut sein

In [31]:
from urllib.parse import urlencode